### PR TITLE
Amd64/hacky code

### PR DESCRIPTION
**Pull request by tkreuzer**
* [BROWSEUI] Disable broken code on x64
* [EXPLORER] Do not initialize m_Position to -1 to stop crashing on x64

### CODE REVIEW

File: dll/win32/browseui/desktopipc.cpp [hunk 1]
```diff
@@ -347,9 +347,11 @@ static HRESULT ExplorerMessageLoop(IEThreadParamBlock * parameters)
     MSG Msg;
     BOOL Ret;
 
+#ifndef _M_AMD64 // FIXME: BROKEN!
     // Tell the thread ref we are using it.
```
Comment (SergeGautherie) [inline dll/win32/browseui/desktopipc.cpp:351]: Sync adding `#ifndef` before or after comments.
Should it be `_WIN64`, like the 2 others?

File: dll/win32/browseui/desktopipc.cpp [hunk 2]
```diff
@@ -519,8 +523,13 @@ extern "C" void WINAPI SHDestroyIETHREADPARAM(IEThreadParamBlock *param)
         param->offset78->Release();
     if (param->offsetC != NULL)
         param->offsetC->Release();
+    // FIXME: This is completely broken!
+    // And I'm sure that there is a dedicated place
+    // in hell for people writing this kind of code!
+#ifndef _WIN64
     if (param->offsetF8 != NULL)
         param->offsetF8->Release();
+#endif
```
Comment (zefklop) [inline dll/win32/browseui/desktopipc.cpp:532]: How is it broken ? From a quick look, I don't see anything wrong.

File: base/shell/explorer/traywnd.cpp [hunk 3]
```diff
@@ -3251,7 +3251,7 @@ class CTrayWindow :
 
     void _Init()
     {
-        m_Position = (DWORD) -1;
```
Comment (zefklop) [inline base/shell/explorer/traywnd.cpp:3254]: What's wrong with this one ?
the constructor already sets this value to 0, and you're effectively changing behaviour where 0 means ABE_LEFT.

See [this](https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/include/psdk/shellapi.h;hb=32b0cf6fc6b8f5fab33c10c782f40d36feebbedb#l17) & [that](https://git.reactos.org/?p=reactos.git;a=blob;f=base/shell/explorer/traywnd.cpp;hb=32b0cf6fc6b8f5fab33c10c782f40d36feebbedb#l2078)

File: dll/win32/browseui/desktopipc.cpp (hunk 2 above)
Comment (tkreuzer) [reply dll/win32/browseui/desktopipc.cpp:532]: An interface pointer called `offsetF8` is wrong enough. Anyway, I debugged what is really wrong and fixed that.

File: dll/win32/browseui/desktopipc.cpp [hunk 4]
```diff
@@ -454,7 +454,7 @@ extern "C" IEThreadParamBlock *WINAPI SHCreateIETHREADPARAM(
 
     TRACE("SHCreateIETHREADPARAM\n");
 
-    result = (IEThreadParamBlock *) LocalAlloc(LMEM_ZEROINIT, 256);
```
Comment (tkreuzer) [inline dll/win32/browseui/desktopipc.cpp:457]: That is one for the hall of shame.

File: dll/win32/browseui/desktopipc.cpp [hunk 5]
```diff
@@ -477,10 +477,10 @@ extern "C" IEThreadParamBlock *WINAPI SHCloneIETHREADPARAM(IEThreadParamBlock *p
 
     TRACE("SHCloneIETHREADPARAM\n");
 
-    result = (IEThreadParamBlock *) LocalAlloc(LMEM_FIXED, 256);
+    result = (IEThreadParamBlock *) LocalAlloc(LMEM_FIXED, sizeof(*result));
     if (result == NULL)
         return NULL;
-    memcpy(result, param, 0x40 * 4);
```
Comment (tkreuzer) [inline dll/win32/browseui/desktopipc.cpp:483]: Yikes! 
Btw, if this was proper C++, the whole function could be `return new IEThreadParamBlock(param);`

Comment (zefklop) [reply dll/win32/browseui/desktopipc.cpp:483]: :vomiting_face:

File: base/shell/explorer/traywnd.cpp (hunk 3 above)
Comment (tkreuzer) [reply base/shell/explorer/traywnd.cpp:3254]: The constructor initializes it to 0, but CreateTrayWindow then calls _Init, setting it to -1.
I get the following crash.
```
kd> kp
 # Child-SP          RetAddr           Call Site
00 00000000`0012e2f0 00000001`4002ec3b explorer!CTrayWindow::ChangingWinPos(struct _WINDOWPOS * pwp = 0x00000000`0012e8a8)+0x320 [E:\amd64\reactos\base\shell\explorer\traywnd.cpp @ 1381] 
01 00000000`0012e380 00000001`4003029f explorer!CTrayWindow::OnWindowPosChanging(unsigned int uMsg = 0x46, unsigned int64 wParam = 0, int64 lParam = 0n1239208, int * bHandled = 0x00000000`0012e3e4)+0x3b [E:\amd64\reactos\base\shell\explorer\traywnd.cpp @ 2480] 
02 00000000`0012e3b0 00000001`40013da1 explorer!CTrayWindow::ProcessWindowMessage(struct HWND__ * hWnd = 0x00000000`000600c6, unsigned int uMsg = 0x46, unsigned int64 wParam = 0, int64 lParam = 0n1239208, int64 * lResult = 0x00000000`0012e528, unsigned long dwMsgMapID = 0)+0x7ef [E:\amd64\reactos\base\shell\explorer\traywnd.cpp @ 3129] 
03 00000000`0012e480 000007ff`b258eb86 explorer!ATL::CWindowImplBaseT<ATL::CWindow,ATL::CWinTraits<1442840576,0> >::WindowProc(struct HWND__ * hWnd = 0x00000000`000600c6, unsigned int uMsg = 0x46, unsigned int64 wParam = 0, int64 lParam = 0n1239208)+0x121 [E:\amd64\reactos\sdk\lib\atl\atlwin.h @ 1594] 
04 00000000`0012e560 fffff800`00405a50 user32!IntCallWindowProcW(int IsAnsiProc = 0n0, <function> * WndProc = 0x00000000`007e0000, struct _WND * pWnd = 0x00000000`004f9c90, struct HWND__ * hWnd = 0x00000000`000600c6, unsigned int Msg = 0x46, unsigned int64 wParam = 0, int64 lParam = 0n1239208)+0x7e6 [E:\amd64\reactos\win32ss\user\user32\windows\message.c @ 1547] 
05 fffff880`044e4690 fffff880`044e4690 nt!KiCallUserMode
```

Comment (tkreuzer) [reply base/shell/explorer/traywnd.cpp]: I added a check for -1 instead. This also fixes the issue.

Comment (learn-more) [reply base/shell/explorer/traywnd.cpp]: Nice.
The -1 is more or less a placeholder for 'Should not use this yet'.
Due to the message passing order this now breaks, ideally all access should use one common accessor that checks for `-1`.
This sounds like a nice starter-task for jira or so.

**Review by zefklop — APPROVED** [review]

**Pull request merged**